sway/tree/view.c

```diff
@@ -475,3 +493,127 @@ void view_child_destroy(struct sway_view_child *child) {
 		free(child);
 	}
 }
+
+static char *parse_title_format(struct sway_view *view) {
+	if (!view->title_format || strcmp(view->title_format, "%title") == 0) {
+		return strdup(view_get_title(view));
+	}
+	const char *title = view_get_title(view);
+	const char *class = view_get_class(view);
+	const char *instance = view_get_instance(view);
+	const char *shell = view_get_type(view);
+	size_t title_len = title ? strlen(title) : 0;
+	size_t class_len = class ? strlen(class) : 0;
+	size_t instance_len = instance ? strlen(instance) : 0;
+	size_t shell_len = shell ? strlen(shell) : 0;
+
+	// First, determine the length
+	size_t len = 0;
+	char *format = view->title_format;
+	char *next = strchr(format, '%');
+	while (next) {
+		len += next - format;
+		format = next;
+
+		if (strncmp(next, "%title", 6) == 0) {
+			len += title_len;
+			format += 6;
+		} else if (strncmp(next, "%class", 6) == 0) {
+			len += class_len;
+			format += 6;
+		} else if (strncmp(next, "%instance", 9) == 0) {
+			len += instance_len;
+			format += 9;
+		} else if (strncmp(next, "%shell", 6) == 0) {
+			len += shell_len;
+			format += 6;
+		} else {
+			++format;
+			++len;
+		}
+		next = strchr(format, '%');
+	}
+	len += strlen(format);
+
+	char *buffer = calloc(len + 1, 1);
+	if (!sway_assert(buffer, "Unable to allocate title string")) {
+		return NULL;
+	}
+
+	// Now build the title
+	format = view->title_format;
+	next = strchr(format, '%');
+	while (next) {
+		// Copy everything up to the %
+		strncat(buffer, format, next - format);
+		format = next;
+
+		if (strncmp(next, "%title", 6) == 0) {
```

Comment:
Is it possible to factor this into a function that returns the formatted length and only calls `strcat` if the `buffer` parameter isn't NULL? We should be able to call it first with `buffer` set to NULL, then allocate the buffer, and then call it again. That would de-duplicate this code.